### ISSUE TITLE
Memory fills up while trying to recognize objects in panoramic images. What can be done?

### ISSUE THREAD

**Issue by kbratsy** [ultralytics/ultralytics#5923]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hello,
I want to detect objects in panoramic images. The memory in my computer is 32 GB. When I want to train panoramic images with 6144x512 100 training data, I cannot do it because the memory is full. What can I do? Also, I specified the imgsz parameter as 6144. Would that be the correct approach? If I reduce the batch size, there may be an improvement in terms of memory, but would a low batch size be suitable for the model?

```
from ultralytics import YOLO
model = YOLO("yolov8x.pt")  # load a pretrained model (recommended for training)
results = model.train(data="data.yaml", epochs=10,device=0,batch=8, imgsz=6144)
```

Thank you.

**Comment by glenn-jocher** [comment]
@kbratsy hi,

Processing panoramic images indeed requires high computational resources, especially memory, due to the large image size. That said, the issue of memory running full can be tackled from a few different angles.

1) **Image Size - imgsz parameter**: Yes, you are correct in matching the image size (imgsz) parameter to the actual size of your panoramic images (6144 in your case). However, this could be the primary reason your memory is getting exhausted. A potential solution could be to reduce the 'imgsz' parameter. It doesn't always have to match the actual image dimensions. The model can still learn to detect objects even in smaller images, and the memory requirement would be significantly lower. Experiment with different sizes to find the optimal balance between model accuracy and resource usage. 

2) **Batch Size**: Reducing the batch size is another approach to decrease memory usage. As you suspected, a low batch size might have an impact on the model's ability to generalize across your dataset. So again, you would need to experiment. The smallest viable batch size would depend on the difficulty and diversity of your dataset. If your different classes are evenly represented and not overly complex, you could get by with a smaller batch size. 

3) **Hardware Upgrade**: Lastly, if changes in the image size or batch size don’t lead to satisfactory results or affect the model's performance, you might want to consider upgrading your hardware, specifically, increasing the RAM. 

Thank you very much for your question and good luck with your project.

**Comment by kbratsy** [comment]
Thank you for your answer. If I reduce the image size (imgsz parameter), the image quality will decrease and this is something I do not want because it will reduce my object detection accuracy (Because I want to recognize Traffic Signs and resolution is important in my case). I'm going to lower the Batch Size parameter a little bit. Also when I research, can we use a renderer that loads a series of images at a time during training instead of loading all the data into memory at once? I have not used such a method before but is this possible in yolov8? 

Thanks.

**Comment by glenn-jocher** [comment]
@kbratsy hello, 

Indeed, reducing the image size might affect the quality, but you're correct that it's a trade-off between computational needs and the level of detail required for your specific task, like traffic sign recognition. 

As for your batch size adjustment, this can certainly help to manage the memory usage by lowering the amount of data being loaded into memory at each step. But remember, the batch size can also influence model learning and convergence, so please monitor your model performance closely. 

In response to your question about using a renderer or a data generator that feeds a batch of images at a time rather than loading all images at once - YOLOv8 already works in this way. The DataLoader employed in YOLOv8 training does not load the entire dataset into memory. Instead, it loads data batch by batch, just as you suggest. So, you should be fine on that front.

Please don't hesitate to reach out if you have any further questions. Good luck with your work!

**Comment by kbratsy** [comment]
Hello,
Thanks for your quick answers. I wanted to train on HPC systems to avoid memory and GPU insufficiency with the images I mentioned. I was able to train under the conditions I wanted, but it calculates the validation values as 0. I checked my validation dataset, there is no problem, I will share the code and output below.
```
import time
import multiprocessing
from ultralytics import YOLO
import torch
from ultralytics.yolo.utils import RANK
torch.cuda.empty_cache()
start_time = time.time()

model = YOLO("yolov8n.pt")

results = model.train(data="/panoramic/data.yaml", epochs=15,dynamic=False,cache=False,mosaic=False,copy_paste=False,mixup=False,augment=False,verbose=True,save_period=10, imgsz=6144,device=[0,1], batch=2)

if RANK==-1:
  results = model.val()
if RANK==0:
  model.export()

end_time = time.time()
elapsed_time = end_time - start_time
minutes, seconds = divmod(elapsed_time, 60)
hours, minutes = divmod(minutes, 60)
print(f"Process: {int(hours):02d}:{int(minutes):02d}:{int(seconds):02d}") 
```
My output,
[panoramic.txt](https://github.com/ultralytics/ultralytics/files/13165289/panoramic.txt)

**Comment by glenn-jocher** [comment]
@kbratsy hello,

Thank you for your additional question. If your validation values are being calculated as zero, that usually hints towards an issue in your validation process or the related dataset. Since you have checked and found no apparent issues with your dataset, it could be possible that something isn't quite right with the configuration or process.

First, make sure you have defined your validation dataset correctly in your "data.yaml" file and that the paths to the validation images and labels are correct. 

Next, check if your labels are in the right format. YOLO expects labels in a certain format: class, x_center, y_center, width, height (all normalized w.r.t. image dimensions).

Also, observe if there could be a possibility that the network is not properly learning from the data, thus leading to poor (zero) validation performance. Note that you've turned off some augmentations which can potentially aid in improving the model's learning.

Lastly, using multiple GPUs (device=[0,1]) can sometimes lead to tricky situations due to synchronization and aggregation between the devices. If none of the above seems to be the issue, I would suggest you to try training on a single GPU for debugging purposes.

I hope this points you in the right direction to identify the issue. Let me know if you have further queries.
